### PR TITLE
+str #17108 adds arity-numbers to javadsl for nicer compile errors

### LINKED ISSUE #17108: add arity number to generated JavaDSL factory method names
to avoid humongous error messages when types don’t match

### PR DESCRIPTION
**Pull request by ktoso**
I opted to keep only arity 0/1/2 as overloads - if any of you have strong feelings about having 3-arity also as overload instead of `create3` speak up - reason is that when testing it out just in sample code i found the verboseness less of a pain as I expected - because the left-hand-side of the `create()` contains a huge type anyway, and it includes the number as well.

Resolves #17108

### CODE REVIEW

**Comment by patriknw** [comment]
LGTM

File: akka-stream/src/main/boilerplate/akka/stream/javadsl/GraphCreate.scala.template [hunk 1]
```diff
@@ -18,17 +18,23 @@ trait GraphCreate {
...
 
-  def partial[S1 <: Shape, S <: Shape, M](g1: Graph[S1, M], block: japi.Function2[FlowGraph.Builder[M], S1, S]): Graph[S, M] =
+  def partial[S1 <: Shape, S <: Shape, M](g1: Graph[S1, M],
+      block: japi.Function2[FlowGraph.Builder[M], S1, S]): Graph[S, M] =
     scaladsl.FlowGraph.partial(g1) { b ⇒ s => block.apply(b.asJava, s) }
 
-  [2..21#def closed[[#S1 <: Shape#], [#M1#], M]([#g1: Graph[S1, M1]#], combineMat: japi.Function1[[#M1#], M],
+  def closed[S1 <: Shape, S2 <: Shape, M1, M2, M](g1: Graph[S1, M1], g2: Graph[S2, M2], combineMat: japi.Function2[M1, M2, M],
+      block: japi.Procedure3[FlowGraph.Builder[M], S1, S2]): RunnableFlow[M] =
+    scaladsl.FlowGraph.closed(g1, g2)(combineMat.apply _) { b => (s1, s2) => block.apply(b.asJava, s1, s2) }
```
Comment (rkuhn) [inline akka-stream/src/main/boilerplate/akka/stream/javadsl/GraphCreate.scala.template]: partial2 is missing

File: akka-stream/src/main/boilerplate/akka/stream/javadsl/SourceCreate.scala.template [hunk 2]
```diff
@@ -9,13 +9,34 @@ import akka.stream.scaladsl.JavaConverters._
 
 trait SourceCreate {
 
+  /**
+   * Creates a `Source` by using a `FlowGraph.Builder[Unit]` on a block that expects
+   * a [[FlowGraph.Builder]] and returns an [[Outlet]].
+   */
```
Comment (rkuhn) [inline akka-stream/src/main/boilerplate/akka/stream/javadsl/SourceCreate.scala.template]: Nice boy-scouting! Although I am the original offender, may I ask you to wave your magic wand also on the other templates? Thanks a bunch!

File: akka-stream/src/main/boilerplate/akka/stream/javadsl/SourceCreate.scala.template [hunk 3]
```diff
@@ -9,13 +9,34 @@ import akka.stream.scaladsl.JavaConverters._
...
   def create[T](block: japi.Function[FlowGraph.Builder[Unit], Outlet[T]]): Source[T, Unit] =
     new Source(scaladsl.Source() { b ⇒ block.apply(b.asJava) })
 
-  def create[T, S <: Shape, M](g1: Graph[S, M], block: japi.Function2[FlowGraph.Builder[M], S, Outlet[T]]): Source[T, M] =
+  /**
+   * Creates a `Source` by using a `FlowGraph.Builder[M]` on a block that expects
+   * a [[FlowGraph.Builder]] and 1 graph and then returns an [[Outlet]].
```
Comment (rkuhn) [inline akka-stream/src/main/boilerplate/akka/stream/javadsl/SourceCreate.scala.template]: perhaps mention that the given graphs are imported as with `Builder.graph` and the resulting shapes are passed to the block

**Comment by rkuhn** [comment]
LGTM apart from the minor fixes

File: akka-stream/src/main/boilerplate/akka/stream/javadsl/GraphCreate.scala.template (hunk 1 above)
Comment (ktoso) [inline akka-stream/src/main/boilerplate/akka/stream/javadsl/GraphCreate.scala.template]: ouch thanks!

File: akka-stream/src/main/boilerplate/akka/stream/javadsl/SourceCreate.scala.template (hunk 2 above)
Comment (ktoso) [inline akka-stream/src/main/boilerplate/akka/stream/javadsl/SourceCreate.scala.template]: Yup, adding comments to all templates

**Comment by ktoso** [comment]
Added comments in all templates, though I admit it's pretty eye-numbing so I hope I got it right - (please have a look before I squash)

File: akka-stream/src/main/boilerplate/akka/stream/javadsl/GraphCreate.scala.template [hunk 4]
```diff
@@ -12,23 +12,83 @@ trait GraphCreate {
...
+
+  /**
+   * Creates a new [[Graph]] by importing the given graphs (using `builder.graph()`) and passing their resulting [[Shape]]s
+   * along with the[[FlowGraph.Builder]] to the given create function.
+   *
+   * Partial graphs are allowed to have unconnected ports.
+   */
+  def partial[S1 <: Shape, S2 <: Shape, S <: Shape, M1, M2, M3, M](g1: Graph[S1, M1], g2: Graph[S2, M2], combineMat: japi.Function2[M1, M2, M],
```
Comment (rkuhn) [inline akka-stream/src/main/boilerplate/akka/stream/javadsl/GraphCreate.scala.template]: M3 is extraneous

File: akka-stream/src/main/boilerplate/akka/stream/javadsl/FlowCreate.scala.template [hunk 5]
```diff
@@ -13,13 +13,32 @@ trait FlowCreate {
   import language.implicitConversions
   private implicit def p[A, B](pair: Pair[A, B]): (A, B) = pair.first -> pair.second
 
+  /**
+   * Creates a Flow by passing a [[FlowGraph.Builder]] to the given create function.
```
Comment (rkuhn) [inline akka-stream/src/main/boilerplate/akka/stream/javadsl/FlowCreate.scala.template]: seen on Sources and Sinks: please also mention the meaning of the expected `Pair<Inlet<I>, Outlet<O>>` return value

**Comment by rkuhn** [comment]
Looking very good, but I still found a minor nitpick :-)

File: akka-stream/src/main/boilerplate/akka/stream/javadsl/GraphCreate.scala.template (hunk 4 above)
Comment (ktoso) [inline akka-stream/src/main/boilerplate/akka/stream/javadsl/GraphCreate.scala.template]: ah, yes, thanks!

File: akka-stream/src/main/boilerplate/akka/stream/javadsl/FlowCreate.scala.template (hunk 5 above)
Comment (ktoso) [inline akka-stream/src/main/boilerplate/akka/stream/javadsl/FlowCreate.scala.template]: done

**Comment by ktoso** [comment]
Comments addressed, will merge after validation

File: akka-stream-tests/src/test/java/akka/stream/javadsl/FlexiMergeTest.java [hunk 6]
```diff
@@ -7,6 +7,7 @@
 import java.util.List;
 import java.util.HashSet;
 
+import akka.stream.javadsl.japi.Function;
```
Comment (drewhk) [inline akka-stream-tests/src/test/java/akka/stream/javadsl/FlexiMergeTest.java]: is this needed? Assuming that the tests were working before, this is an unused import.

**Comment by drewhk** [comment]
LGTM, nice work!

**Comment by rkuhn** [comment]
merge?

**Comment by ktoso** [comment]
Yeah :-)

**Pull request merged**